### PR TITLE
Enable multiple strategies

### PR DESCRIPTION
**Pull request by maxim-belkin**
- Allow multiple strategies to match URL
- factor out available strategies into a method
- `strat_matcher` method returns whether specified strategy is applicable to specified url
- warn when no strategies have been matched
- guard `Matched Text on Page` and `Matched Versions` outputs by `Homebrew.args.verbose?`

Formulae that can be used to see the effect of this PR: xmp, wtf, aespipe
 
<details>
<summary><code>brew livecheck -d xmp</code></summary>
<pre><code>
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/brew.rb (Formulary::TapLoader): loading /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/xmp.rb
Formula:         xmp
Livecheckable?:  No

URL:             https://git.code.sf.net/p/xmp/xmp-cli.git
\# of strategies: 2
Strategies:      git, sourceforge
Using strategy: git
Found new versions: 2.2.0, 2.3.0, 2.3.1, 2.3.2, 2.4.0, 2.4.1, 2.5.0, 2.5.1, 2.7.0, 2.7.0-win32-amigaconf.patch, 2.7.1, 3.0.0, 3.0.0-pre1, 3.0.1, 3.1.0, 3.2.0, 3.3.0, 3.4.0, 3.4.1, 3.5.0, android-1.0.1, android-1.1, android-1.2, android-1.9, android-1.9.1, android-1.9.2, android-1.9.3, android-1.9.4, android-1.9.5, android-2.0.0, android-2.0.1, android-2.0.2, android-2.0.3, android-2.1.0, android-2.1.1, android-2.2.0, android-2.3.0, android-2.4.0, android-2.9.0, android-2.9.1, android-2.9.2, android-3.0.0, android-3.0.1, android-3.0.2, android-3.1.0, android-3.1.1, android-3.1.2, android-3.2.0, android-3.2.1, android-3.3.0, android-3.4.0, android-3.4.1, android-3.4.2, android-3.4.3, android-3.4.4, api-4.0, libxmp-3.9.0, libxmp-3.9.1, libxmp-3.9.2, libxmp-3.9.3, libxmp-3.9.4, libxmp-4.0.0, libxmp-4.0.1, libxmp-4.0.2, libxmp-4.0.3, libxmp-4.0.4, libxmp-4.1.0, libxmp-4.1.1, libxmp-4.1.2, libxmp-4.1.3, libxmp-4.1.4, r2_2_0, r2_2_0_pre1, r2_2_0_pre2, r2_2_0_pre3, r2_2_0_pre4, r2_2_0_pre5, r2_2_0_pre6, r2_3_0, r2_3_0_pre1, r2_3_0_pre2, r2_3_0_pre3, r2_3_1, r2_3_2, r2_4_0, r2_4_0_pre1, r2_4_0_pre2, r2_4_0_pre3, r2_4_0_pre4, r2_4_0_pre5, r2_4_0_pre6, r2_4_1, r2_5_0, r2_5_0_pre1, r2_5_0_pre3, r2_5_0_pre4, r2_5_0_pre5, r2_5_0_pre6, r2_5_1, v2.6.0, v2.6.1, v2.6.2, xmp-3.9.0, xmp-3.9.1, xmp-4.0.0, xmp-4.0.1, xmp-4.0.10, xmp-4.0.11, xmp-4.0.2, xmp-4.0.3, xmp-4.0.4, xmp-4.0.5, xmp-4.0.6, xmp-4.0.7, xmp-4.0.8, xmp-4.0.9, xmp-4.1.0
Using strategy: sourceforge
Found new versions: 4.4.1, 4.1.0, 4.4.0, 4.3.13, 4.0.11, 4.3.12, 4.3.11, 4.3.10, 4.3.9, 4.3.8, 4.3.7, 4.3.6, 4.3.5, 4.3.4, 4.3.3, 4.0.10, 4.3.2, 4.3.1, 4.3.0, 4.0.9, 4.0.8, 4.2.8, 4.2.7, 4.2.6, 4.2.5, 4.0.7, 4.2.4, 4.2.3, 4.2.2, 4.2.1, 4.2.0, 4.1.5, 4.0.6, 4.1.4, 4.0.5, 4.1.3, 4.1.2, 4.1.1, 4.0.4, 4.0.3
xmp : 4.1.0 ==> 4.4.1</code></pre>
</details>

<details>
<summary><code>brew livecheck -d wtf</code></summary>
<pre><code>
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/wtf.rb
Loading /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-livecheck/Livecheckables/wtf.rb
Formula:         wtf
Livecheckable?:  Yes

URL:             https://sourceforge.net/projects/bsdwtf/
\# of strategies: 2
Strategies:      sourceforge, page_match
Regex:           /.*?\/wtf-([0-9.]+)\.t/
Using strategy: sourceforge
Found new versions: 20200613, 20200611, 20200610, 20200609, 20200606, 20200605, 20200604, 20200603, 20200518, 20200505, 20200429, 20200427, 20200422, 20200326, 20200226, 20200217, 20200208, 20200130, 20200127, 20200110, 20191223, 20191214, 20191201, 20191111, 20191020, 20191006, 20191002, 20190925, 20190909, 20190903, 20190902, 20190901, 20190830, 20190822, 20190817, 20190728, 20190726, 20190723, 20190720, 20190711, 20190705, 20190704, 20190703, 20190702, 20190630, 20190629, 20190628, 20190623, 20190622, 20190621, 20190617, 20190612, 20190611, 20190608, 20190607, 20190604, 20190602, 20190527, 20190524, 20190516, 20190507, 20190424, 20190422, 20190417, 20190410, 20190405, 20190402, 20190401, 20190330, 20190323, 20190308, 20190304, 20190224, 20190221, 20190219, 20190201, 20190128, 20190123, 20190122, 20190114, 20190107, 20181230, 20181212, 20181207, 20181205, 20181119, 20180925, 20180919, 20180816, 20180813, 20180704, 20180627, 20180624, 20180623, 20180621, 20180620, 20180615, 20180529, 20180525, 20180522
Using strategy: page_match
No new versions found using page_match
wtf : 20200613 ==> 20200613</code></pre>
</details>

<details>
<summary><code>brew livecheck -d aespipe</code></summary>
<pre><code>
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/aespipe.rb
Formula:         aespipe
Livecheckable?:  No

URL:             https://loop-aes.sourceforge.io/aespipe/aespipe-v2.4f.tar.bz2
\# of strategies: 0
Strategies:      None found
Warning: No strategies identified for aespipe using URL: https://loop-aes.sourceforge.io/aespipe/aespipe-v2.4f.tar.bz2

URL:             https://loop-aes.sourceforge.io/
\# of strategies: 0
Strategies:      None found
Warning: No strategies identified for aespipe using URL: https://loop-aes.sourceforge.io/
Error: aespipe: Unable to get versions</code></pre>
</details>

### CODE REVIEW

**Comment by samford** [comment]
## Discussion

As mentioned in the last PR, it would be helpful to have a heads up when you plan to work on something big like this before it becomes a PR. This is especially important right now because I'm making significant internal changes to livecheck and it's not visible until I create PRs.

The only way to know what's going on is to discuss it, so we need a line of communication. I'm going to have an admin add to you to the #livecheck channel on Slack but also feel free to DM me. I'm happy to discuss changes like this and how they fit within what's happening in livecheck right now, as it will help us both to avoid wasting time/effort.

## Duplicated Effort

Most of the changes here are already covered in my work on the forthcoming `strategy` feature. This work has been done for well over a month now and I've been creating incremental PRs as time permits (our GSoC project is very time-consuming, as the ongoing flood of PRs may suggest).

The next PR I'll be creating is a significant refactoring of how the strategies are created/organized and achieves the same goals as some of your changes here. This leads up to the strategy feature itself, which also impacts this PR.

## Benefits vs. Detriments

That said, I don't currently see the benefits of allowing livecheck to execute multiple strategies for a given formula. I do see some notable detriments that would suggest it's not something we want to do.

Without a livecheckable, livecheck currently tries URLs from the formula and stops when it finds a version. This works fine and when it doesn't it's arguably a _good_ thing that livecheck fails, as it lets us know which formulae need livecheckable additions/modifications.

Even if trying multiple strategies could lead to finding the newest version when trying one strategy wouldn't, it's not something that I think is appropriate. This could easily lead to a situation where a "guessed" check is finding the correct version but it's unnecessarily checking several strategies in the process. Since the check finds the latest version, it's unlikely that anyone would write a livecheckable for it and we would be stuck with a default check that does more work than it should. This is currently true with URLs for some of the "guessed" checks and doing the same thing for strategies would make it worse.

I feel that it's appropriate to draw the line at executing one strategy per URL and if it doesn't work correctly we simply create a livecheckable of some sort. If we allow a check to potentially execute more than one strategy, we could end up in the worst case scenario where we're running multiple strategies on multiple URLs and this can become a big issue when you're working with ~3000 guessed checks.

Besides this, another thing to note is that there are times where, for example, a Git repo's tags cause livecheck to return something other than the latest version for the formula. Running multiple strategies for a given formula could lead to a situation where source A has the correct latest version but source B has a bad version that's treated as newest but isn't really. In this case, using two strategies would seemingly give us a bad version where using one strategy wouldn't.

## Extracting Changes

The only thing that I could see moving forward here is some of the changes to the debug output. That is, there may be some value in using the comma-separated style for the versions unless the `--verbose` flag is present.

However, I definitely wouldn't want to get rid of the format where we print the parsed `Version` object, as that's very valuable for me on a regular basis. We could simply use the current format for `--verbose`, as the "verbose" style in this PR isn't useful (it will almost always display duplicate information like `Match 94: 20180621 => 20180621`, which isn't valuable).

Similarly, moving the "matched text on page" for the `page_match` strategy behind the `--verbose` flag may be good as well.

That said, I would want to test the debug output changes for a bit and get feedback from the other regular livecheck users to see how we all feel about it. The purpose of the `--debug` flag is to give the user an idea of how livecheck is operating underneath the hood and I'm wary of reducing the default information.

## Summary

Long story short, I don't see a way forward for this PR. Most of these changes are already being handled in a different way (my work on the `strategy` feature). I'm having trouble seeing any benefits here whereas the detriments stand out strongly to me, so I don't think this is the right path to take.

I would have preferred to have discussed this beforehand, to save you the effort. I appreciate that you want to improve livecheck and are willing to put in the effort but we need to discuss and coordinate our work, especially right now.

**Comment by samford** [comment]
I took some time today to rebase my strategy refactoring work, put some finishing touches on it, and create a PR for it (#1025). As you can see, it covers a lot of the changes here, outside of the ones that are related to enabling multiple strategies and the aforementioned changes to the "Matched Version" debug output. I would appreciate your input, if you have the time.

**Comment by maxim-belkin** [comment]
Thanks for the detailed response, Sam. Please see my responses below.

Maxim

> I don't currently see the benefits of allowing livecheck to execute multiple strategies for a given formula.

I didn't either until I did it. It can be an option rather than the default behavior but there is definitely benefit to at least enabling that.

>  took some time today to rebase my strategy refactoring work, put some finishing touches on it, and create a PR for it (#1025). As you can see, it covers a lot of the changes here, outside of the ones that are related to enabling multiple strategies and the aforementioned changes to the "Matched Version" debug output. I would appreciate your input, if you have the time.

I'll happily review it. I'll try to do it tonight or over the weekend.

> As mentioned in the last PR, it would be helpful to have a heads up when you plan to work on something big like this before it becomes a PR. This is especially important right now because I'm making significant internal changes to livecheck and it's not visible until I create PRs.

Homebrew is an open-source project on GitHub. All contributions must be submitted in the form of a PR. Including those from maintainers. Until a PR is created, changes don't exist. This is a general consensus. Otherwise you'd be making yourself a privileged contributor and that can be considered discriminatory toward regular contributors. Homebrew's PR template has a line that asks contributors to check that there aren't other open pull requests for the same change. There is no line that says "Check in with maintainers to see if they're working on something similar".

This is not the first time you mention your changes that (a) I don't have access to (b) not submitted in the form of a PR as a reason for not wanting to approve my PR. I welcome and am responsive to constructive criticism but I'm not sure how anyone can respond to "I've been working on something similar but it's not there yet". If it's not ready -- it's not ready. Maintainers have two independent "responsibilities": 1. reviewing PRs from contributors ("Maintainer"), and 2. working on improvements themselves ("Developer"). If you prioritize the latter, you'll surely do what you think is right. But the project will quickly become stale (and might even die off if no one steps in) because this is not your main job and you can't devote all of you time to it. That's the idea behind open-source -- everyone chips in a little bit of their time and expertise when they can and feel like doing so. I work on my contributions as a  "warm up" for my daily job or when I need to "cool off" (I started writing this response instead of my coding "warm up" and now I'm finishing it up to "cool off").


> Most of the changes here are already covered in my work on the forthcoming `strategy` feature. This work has been done for well over a month now and I've been creating incremental PRs as time permits (our GSoC project is very time-consuming, as the ongoing flood of PRs may suggest).
> 
> The next PR I'll be creating is a significant refactoring of how the strategies are created/organized and achieves the same goals as some of your changes here. This leads up to the strategy feature itself, which also impacts this PR.

PR #601 was submitted on May 7, a week and a half short of two months ago and there is a lot in common with what you've got. I'm not going to argue about the differences and what should or shouldn't have been done and when because (a) I see the differences and (b) it's not going to change anything.

The main message I want you to get is that maintainers are not (only) developers. You show excellent developer skills but need to improve on those of a maintainer. This is your opportunity to learn how to manage people and their time. It's an important job within the broad open-source community. Contributors have to be encouraged and their efforts appreciated (especially when they're not being incorporated).

I see your good intentions (you tweaked my previous PR to your liking, you wrote long reviews) and I see that you've got the necessary programming skills... but I also see some issues with how you manage this repo that I tried to outline above. I think it would be best if Homebrew can organize on-boarding/best practices sessions for maintainers to avoid situations / conversations like this.

With that, I think it makes sense to close this PR. I'm still going to try to convince you that allowing multiple strategies is a good idea but in a separate PR.

Thank you for all you do for the project and this repo! And good luck with GSoC!
